src/sync/partial_sync.cpp

```diff
@@ -16,20 +16,109 @@
 //
 ////////////////////////////////////////////////////////////////////////////
 
-#include "sync/partial_sync.hpp"
-
-#include "impl/notification_wrapper.hpp"
-#include "impl/object_accessor_impl.hpp"
 #include "object_schema.hpp"
 #include "results.hpp"
 #include "shared_realm.hpp"
+#include "subscription_state.hpp"
+#include "impl/collection_change_builder.hpp"
+#include "impl/notification_wrapper.hpp"
+#include "impl/object_accessor_impl.hpp"
+#include "sync/partial_sync.hpp"
 #include "sync/sync_config.hpp"
+#include "sync/sync_session.hpp"
 
 #include <realm/util/scope_exit.hpp>
 
 namespace realm {
 namespace partial_sync {
 
+SubscriptionState create_or_update_subscription(Realm::Config config, SharedGroup &sg, realm::_impl::CollectionChangeBuilder &changes, Query &query, SubscriptionState previous_state) {
+
+    if (!config.sync_config || !config.sync_config->is_partial) {
+        return SubscriptionState::NOT_SUPPORTED;
+    }
+
+    // FIXME: Question: Is it problematic to do a write transaction here? Should we move it to a background thread`?
```

Comment:
Yes, it is very problematic. This design does not work at all outside of the most trivial cases. The RealmCoordinator owns the SharedGroup and very carefully manages what version it is at. Changing that version will break everything.

Reply:
Browsing through the code, it appeared the SharedGroup wasn't used anymore after this step in `calculate_changes`, but I could, of course, have missed something?

It should be possible to move this write to another thread, and then optimistically assume it was successful. It will introduce a race condition where you might get an initial callback followed by an error, but since partial sync errors are asynchronous anyway, that is probably acceptable. 

In Java, I would use something like a thread pool to prevent spawning threads all over the place, but I don't know if we have something similar here that can be used?

Reply:
A single SharedGroup instance is used for every run of all of the notifiers for a given realm path.

I don't understand why the subscription is being created in ResultsNotifier::calculate_changes() in the first place. It makes sense for the subscription status to be read there, but creating it is entirely unrelated to anything that ResultsNotifier does. I would expect the subscription to be created when the notification is registered on the collection.

Reply:
Yes, that would be a more natural place to create it, the problem is that any errors should still come through the notification system, but perhaps I can record them somewhere and just throw an exception as soon as the notification trigger similar to what is being done if something is wrong when the Relm is being opened.

I'll look into it.

Reply:
Actually, now I remember why. Originally I discarded the idea because it would imply writing to the Realm registering the callback which would advance it (which would be bad). But maybe we can work around that by opening a specific shared group. 

The downside of that approach is that it means that every time an observer is registered you would have to open a shared group and close it again, but that might be cheap enough?

Reply:
So, talking to @finnschiermer, SharedGroups are pretty expensive and also have an impact on file descriptors, so if we have an extra SharedGroup for each thread that can register an observer it will be quite some overhead

I see two options here really (might be more):

1) Keep the write in the notifier thread and add an extra SharedGroup for writing only (cleaner code/abstraction)
2) Maintain an extra SharedGroup pr. thread that uses callbacks  in order to not accidentially advance the Realm when registering the callback

What are your thoughts @tgoyne ?

Reply:
An entirely untested and unpolished example of the sort of design I have in mind: https://github.com/realm/realm-object-store/commit/12dc158be5efbc7e499f36f564fec1cb314e657b

Reply:
In case it is not clear, I think trying to make writes of any kind from a notifier (no matter how indirectly) is a very very bad idea. The notifier lifecycle and functionality is already far too complicated even with them being purely read-only, and extending them to also make writes to the Realm will make them even more nightmarish.

Reply:
Thanks for that commit 👍 
I did consider a similar approach but discarded it mostly because it introduces some rather annoying race conditions regarding error handling. 

Most noticeably that you optimistically have to assume that creating the subscription will succeed thus you risk having situations where the first callback returns a query and the next callback returns an error saying the subscription could not be created.

However, since Partial Sync query validation is async as well, checking for name collision that way might be acceptable. Also, we can still do a synchronous read to check before even notifying the partial sync worker thread, which would drastically limit the number of times we would have to use async error handling.

And I agree that the approach you outlined has a much cleaner architecture, so I don't see a problem going with that for a first implementation.